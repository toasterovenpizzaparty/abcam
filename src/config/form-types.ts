// These steps enums serve as keys for our step object within our state.
export enum STEPS {
  STEP_ONE = "step-one",
  STEP_TWO = "step-two",
}

// These fields enums serve as keys for our fields within our step object.
export enum FIELDS {
  RATING = "rating",
  TITLE = "title",
  DESCRIPTION = "description",
  IMAGE = "image",
  IMAGE_DESCRIPTION = "image-description",
}
