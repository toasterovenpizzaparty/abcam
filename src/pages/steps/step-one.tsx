import React from "react";

export const StepOne: React.FC = () => {
  return <h1>Hello, world!</h1>;
};
