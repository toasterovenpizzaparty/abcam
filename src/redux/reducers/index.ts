export {
  StepReducer,
  ActionTypes as StepActionTypes,
} from "./step-reducer/step-reducer";
export type { State as StepState } from "./step-reducer/step-reducer";
export {
  APIReducer,
  ActionTypes as APIActionTypes,
} from "./api-reducer/api-reducer";
export type { State as APIState } from "./api-reducer/api-reducer";
