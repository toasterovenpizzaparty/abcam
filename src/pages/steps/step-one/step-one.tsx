import React, { useCallback } from "react";
// import { useHistory } from "react-router-dom";
import { Rating, TextField } from "../../../components";
import { FIELDS, STEPS } from "../../../config/form-types";
// import { ServerResponse } from "../../redux/reducers/step-reducer";
import { useStepSelector, createUseFieldState } from "../../../redux/selectors";
// import { useCreateReviewAction } from "../../redux/actions";
import { StepPropTypes } from "../../../types/types";

export const StepOne: React.FC<StepPropTypes> = ({
  id = "step-one",
  isError = false,
  errorMessage = "Oh no an error has occured trying to submit your review.",
}) => {
  const useFieldState = createUseFieldState(id);
  const [title, setTitle] = useFieldState({
    fieldKey: FIELDS.TITLE,
  });
  const [rating, setRating] = useFieldState({
    fieldKey: FIELDS.RATING,
    defaultValue: "-1",
  });
  return (
    <aside data-test-id='step-one'>
      {isError && <span data-test-id='error-message'>{errorMessage}</span>}
      <Rating
        label='Rating'
        data-test-id='rating'
        selectedRating={rating}
        onRatingChange={setRating}
      />
      <TextField
        label='Title'
        placeholder='a great titel e.g. a once in a lifetime product'
        value={title}
        onTextChange={setTitle}
        data-test-id='title'
      />
    </aside>
  );
};
