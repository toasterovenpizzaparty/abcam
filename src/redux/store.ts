import {
  createStore,
  applyMiddleware,
  compose,
  combineReducers,
  StoreEnhancer,
} from "redux";
import thunk from "redux-thunk";
import { StepReducer, StepState, APIReducer, APIState } from "./reducers";
const initialState = {};
const enhancers: StoreEnhancer[] = [];
const middleware = [thunk];

export type RootState = {
  steps: StepState;
  api: APIState;
};

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  compose(...enhancers)
);
const combinedReducers = combineReducers({
  steps: StepReducer,
  api: APIReducer,
});

const getStore = () =>
  createStore(combinedReducers, initialState, composedEnhancers);
const store = getStore();

export default store;
export { getStore };
