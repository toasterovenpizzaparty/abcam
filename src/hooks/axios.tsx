import React, { useState, useEffect, useContext, useCallback } from "react";
import { AxiosRequestConfig, AxiosResponse, AxiosError } from "axios";
import { AxiosContext } from "../providers/axios";

type AxiosState = {
  isLoading: boolean;
  data?: AxiosResponse;
  error?: AxiosError;
};

const mergeConfigurations = (
  configurationA: AxiosRequestConfig,
  configurationB?: AxiosRequestConfig
) => {
  if (configurationA && configurationB) {
    return { ...configurationA, ...configurationB };
  }
  return configurationA;
};

export const useAxios = () => useContext(AxiosContext);

export const useAxiosCache = (config: AxiosRequestConfig) => {
  const { client } = useContext(AxiosContext);
  const [state, setState] = useState<AxiosState>({
    isLoading: false,
  });

  const fetchData = useCallback(async (optionalConfig?: AxiosRequestConfig) => {
    setState((state) => ({ ...state, isLoading: true }));
    try {
      const data = await client.request(
        mergeConfigurations(config, optionalConfig)
      );
      setState({ data, isLoading: false });
      return data;
    } catch (error) {
      setState({ error, isLoading: false });
      return error;
    }
  }, []);

  return { ...state, fetchData };
};

// Creates a hook that pulls in the API Response upon mount.
/* export const useAxiosMount = (config: AxiosRequestConfig) => {
  const axios = useAxios(config);

  useEffect(() => {
    axios.fetchData();
  }, []);

  return axios;
}; */
