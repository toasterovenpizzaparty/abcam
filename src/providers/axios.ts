import axios from "axios";

export const client = () => {
  return axios.create({
    baseURL: "https://private-8c57ff-stepbystep.apiary-mock.com",
    timeout: 1000,
  });
};
