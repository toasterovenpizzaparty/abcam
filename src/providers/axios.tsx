import React, { createContext } from "react";
import axios, { AxiosInstance } from "axios";

export const client = () => {
  return axios.create({
    baseURL: "https://private-8c57ff-stepbystep.apiary-mock.com",
    timeout: 1000,
  });
};

export const AxiosContext = createContext({ client: client() });

export const AxiosProvider: React.FC<
  React.Provider<{ client: AxiosInstance }>
> = ({ children }) => (
  <AxiosContext.Provider value={{ client: client() }}>
    {children}
  </AxiosContext.Provider>
);
