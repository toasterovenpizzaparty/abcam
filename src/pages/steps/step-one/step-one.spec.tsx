import React from "react";
import axios from "axios";
import { mount, shallow } from "enzyme";
import { Provider } from "react-redux";
import { getStore } from "../../../redux/store";
import "../../../config/enzyme";
import { Router } from "react-router-dom";
import { StepOne } from "./step-one";
import { STEPS, FIELDS } from "../../../config/form-types";
import { ActionTypes } from "../../../redux/reducers";

const mockedAxios = axios as jest.Mocked<typeof axios>;
jest.mock("axios");

describe("<Step One />", () => {
  it("Renders the page", () => {
    const wrapper = shallow(
      <Provider store={getStore()}>
        <StepOne />
      </Provider>
    ).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it("Saves the users input data in the store", () => {
    const reduxStore = getStore();
    const wrapper = mount(
      <Provider store={reduxStore} data-test-id='redux-provider'>
        <StepOne />
      </Provider>
    );
    wrapper.find('[data-test-id="title"]').simulate("change", {
      target: {
        value: "foobar",
      },
    });
    wrapper.update();

    expect(wrapper.find('[data-test-id="title"]').props().value).toEqual(
      "foobar"
    );
    wrapper
      .find('[data-test-id="rating"] input')
      .at(2)
      .simulate("change", { target: { checked: true } });
    wrapper.update();

    expect(
      wrapper.find('[data-test-id="rating"]').props().selectedRating
    ).toEqual("3");

    console.log(reduxStore.getState());

    expect(reduxStore.getState()).toEqual({
      steps: {
        [STEPS.STEP_ONE]: {
          fields: { title: "foobar", rating: "3" },
        },
      },
    });
  });

  it("disabled the submit if API is loading or insufficient input is given", () => {
    const reduxStore = getStore();
    const wrapper = mount(
      <Provider store={reduxStore} data-test-id='redux-provider'>
        <StepOne />
      </Provider>
    );

    expect(
      wrapper.find('[data-test-id="submit-button"]').props().disabled
    ).toBe(true);

    reduxStore.dispatch({
      type: ActionTypes.SAVE_FORM_FIELD,
      stepKey: STEPS.STEP_ONE,
      fields: {
        formFieldKey: FIELDS.RATING,
        value: "2",
      },
    });
    reduxStore.dispatch({
      type: ActionTypes.SAVE_FORM_FIELD,
      stepKey: STEPS.STEP_ONE,
      fields: {
        formFieldKey: FIELDS.TITLE,
        value: "foobar",
      },
    });
    wrapper.update();

    expect(
      wrapper.find('[data-test-id="submit-button"]').props().disabled
    ).toBe(false);

    reduxStore.dispatch({
      type: ActionTypes.START_REQUEST,
      stepKey: STEPS.STEP_ONE,
    });
    wrapper.update();

    expect(
      wrapper.find('[data-test-id="submit-button"]').props().disabled
    ).toBe(true);

    reduxStore.dispatch({
      type: ActionTypes.FINISH_REQUEST,
      stepKey: STEPS.STEP_ONE,
      serverResponse: {
        data: null,
        error: null,
      },
    });
    wrapper.update();

    expect(
      wrapper.find('[data-test-id="submit-button"]').props().disabled
    ).toBe(false);
  });

  it("shows an error message if the API throws an error", async () => {
    const reduxStore = getStore();
    const wrapper = mount(
      <Provider store={reduxStore} data-test-id='redux-provider'>
        <StepOne />
      </Provider>
    );

    expect(wrapper.find('[data-test-id="error-message"]').exists()).toBe(false);

    reduxStore.dispatch({
      type: ActionTypes.FINISH_REQUEST,
      stepKey: STEPS.STEP_ONE,
      serverResponse: {
        data: null,
        error: { message: "foo bar" },
      },
    });
    wrapper.update();

    expect(wrapper.find('[data-test-id="error-message"]').exists()).toBe(true);
  });

  it("Submits the data to an API and pushes a route with the correct id", async () => {
    const expectedId = 3;
    axios.create.mockImplementation((config) => axios);
    axios.request.mockImplementationOnce(() => ({
      data: {
        id: expectedId,
      },
    }));
    const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };

    const reduxStore = getStore();
    const wrapper = mount(
      <Provider store={reduxStore} data-test-id='redux-provider'>
        <Router history={historyMock}>
          <StepOne />
        </Router>
      </Provider>
    );

    reduxStore.dispatch({
      type: ActionTypes.SAVE_FORM_FIELD,
      stepKey: STEPS.STEP_ONE,
      fields: {
        formFieldKey: FIELDS.RATING,
        value: "2",
      },
    });
    reduxStore.dispatch({
      type: ActionTypes.SAVE_FORM_FIELD,
      stepKey: STEPS.STEP_ONE,
      fields: {
        formFieldKey: FIELDS.TITLE,
        value: "foobar",
      },
    });
    wrapper.update();

    wrapper.find('[data-test-id="submit-button"]').simulate("click");
    wrapper.update();

    // Flush promises
    await new Promise((resolve) => setImmediate(resolve));

    const route = ((historyMock.push.mock.calls || []).pop() || []).pop();
    expect(route).toEqual(`/step/2/id/${expectedId}`);
  });
});
