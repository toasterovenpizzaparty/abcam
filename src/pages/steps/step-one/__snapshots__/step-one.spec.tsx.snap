// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Step One /> Renders the page 1`] = `ShallowWrapper {}`;
